platform/sync-installer/sync-installer-impl/src/main/java/org/codice/ddf/sync/installer/impl/SynchronisedInstallerImpl.java

```diff
@@ -0,0 +1,566 @@
+        servicePid);
+    ServiceConfigurationListener listener = new ServiceConfigurationListener(servicePid);
+    ServiceRegistration<?> registration =
+        bundleContext.registerService(ConfigurationListener.class.getName(), listener, null);
+
+    Callable<Boolean> isServiceUpdated =
+        () -> {
+          LOGGER.debug(
+              "Waiting for service with pid [{}] to reflect configuration updates...", servicePid);
+          return listener.isUpdated();
+        };
+
+    boolean available;
+
+    LOGGER.info("Updating configuration of service with pid [{}].", servicePid);
+
+    try {
+      Configuration config = configAdmin.getConfiguration(servicePid, bundleLocation);
+      config.setBundleLocation(bundleLocation);
+      config.update(toDic(properties));
+
+      LOGGER.debug("Updated configuration of service with pid [{}].", servicePid);
+      available = wait(isServiceUpdated, maxWaitTime, DEFAULT_POLLING_INTERVAL);
+    } catch (IOException e) {
+      throw new SynchronisedInstallerException(
+          "Failed to update managed service configuration with pid of [" + servicePid + "]", e);
+    } catch (Exception e) {
+      throw new SynchronisedInstallerException(
+          "Interrupted while waiting for service with pid of [" + servicePid + "] to be updated.",
+          e);
+    } finally {
+      registration.unregister();
+    }
+
+    if (!available) {
+      throw new SynchronisedInstallerException(
+          "Managed service failed to update within [" + maxWaitTime + "] ms.");
+    }
+  }
+
+  @Override
+  public void waitForServiceAvailability(String servicePid) throws SynchronisedInstallerException {
+    waitForServiceAvailability(DEFAULT_MAX_SERVICE_WAIT, servicePid);
+  }
+
+  @Override
+  public void waitForServiceAvailability(long maxWaitTime, String servicePid)
+      throws SynchronisedInstallerException {
+    ServiceTracker st = getServiceTracker(bundleContext, getServiceIdFilter(servicePid));
+    st.open(true);
+
+    Callable<Boolean> isServiceAvailable =
+        () -> {
+          LOGGER.debug("Waiting for service with pid [{}] to be registered", servicePid);
+          return st.getService() != null;
+        };
+
+    try {
+      wait(
+          isServiceAvailable,
+          maxWaitTime,
+          DEFAULT_POLLING_INTERVAL,
+          "Managed service failed to appear after [" + maxWaitTime + "] ms.");
+    } finally {
+      st.close();
+    }
+  }
+
+  @Override
+  public void installFeatures(String feature, String... additionalFeatures)
+      throws SynchronisedInstallerException {
+    installFeatures(DEFAULT_MAX_FEATURE_WAIT, feature, additionalFeatures);
+  }
+
+  @Override
+  public void installFeatures(long maxWaitTime, String feature, String... additionalFeatures)
+      throws SynchronisedInstallerException {
+
+    Set<String> featuresToInstall =
+        getFeaturesFromNames(feature, additionalFeatures)
+            .filter(f -> !featuresService.isInstalled(f))
```

Comment:
❗️ I don't think you need to worry about those already installed, let the featureService handle that properly.

Reply:
what happens if the system changes in between your check and the call to install?